### ISSUE TITLE
renderer call chain fails with undefined on custom extension

### ISSUE THREAD

**Issue by codeedog** [markedjs/marked#2151]
**Marked version:** 2.1.3

**Describe the bug**
Using a custom extension (e.g. `media`) with a `renderer`, I wanted to allow overriding the default behavior. In the default `renderer`, I test for `this.renderer.media` and call it if it exists. If it returns `false` or doesn't exist, I call my default renderer. `marked.use({ renderer: { media(...) {}}})` creates and overrides the `media` extension. When the entire chain returns `false`, node errors with `Cannot read property 'apply' of undefined`.

**To Reproduce**

*Bug involves .use() extension; cannot replicate in online system.*

Run the following code:

````
// Sample extension
const md = '@[alternative](source)';

// Create Media extention that mimics image with '@' instead of '!'
const mediaExtension = {
  name: 'media',
  level: 'inline',
  start(src) { return src.match(/^@\[/)?.index; },
  tokenizer(src, tokens) {
    const rule = new RegExp(/^@\[([^\]]*)\]\(([^\)]+)\)/);
    const match = rule.exec(src);
    if (match) {
      return { type: 'media', raw: match[0], alt: match[1].trim(), src: match[2].trim() };
    }
  },
  renderer(token) {
    let html = false;
    // *** Check for override of 'media' extension. If present, call it.
    if (this.renderer.media) html = this.renderer.media(token.alt, token.src);
    return html !== false ? html : `<media src="${token.src}" alt="${token.alt}"></media>`;
  },
};

// Succeeds
marked.use({extensions: [mediaExtension]});
console.log("Test(clean ext):", marked(md));

let renderInFirstFunc = true;
function firstFunc(alt, src) {
  console.log("Func1:", renderInFirstFunc); 
  return renderInFirstFunc && `<ff>${src}, ${alt}</ff>`;
}

// Succeeds
marked.use({ renderer: { media: firstFunc }});
marked.use({ renderer: { media(alt, src) { console.log("Func2"); return false; }}});
console.log("Test(render #1):", marked(md));

// Fails with error. Falls through renderer call chain in marked.js:212
renderInFirstFunc = false;
console.log("Test(apply nul):", marked(md));
````

**Expected behavior**
If an extension is overridden in renderer, a false should be returned to the extension renderer so that it can fallback to its default behavior.

**Output of above code**

````
Test(clean ext): <p><media src="source" alt="alternative"></media></p>

Func2
Func1: true
Test(render #1): <p><ff>source, alternative</ff></p>

Func2
Func1: false
/Users/andrewphilips/Documents/Websites/websites/node_modules/marked/src/marked.js:121
    throw e;
    ^

TypeError: Cannot read property 'apply' of undefined
Please report this to https://github.com/markedjs/marked.
    at Renderer.renderer.<computed> (/Users/andrewphilips/Documents/Websites/websites/node_modules/marked/src/marked.js:212:32)
````

**Comment by UziTech** [comment]
I think the issue is that there is no `media` renderer so `marked.use({ renderer: { media: ... }});` doesn't really make sense. The `renderer` option is to override existing marked renderers not extensions.

**Comment by codeedog** [comment]
Yes, however, once an extension is created, how may one override it? Perhaps this is a feature request? I don’t mind calling through the override code like I did in the extended renderer function. If the code doesn’t fail, the extension code can recover. It allows the extension code to model the provided (standard) features by just not calling into an undefined. So, bug fix or feature request? I’m happy to file a fix for review.

**Comment by calculuschild** [comment]
You can create multiple extensions with the same name, and the latest one loaded will override the earlier ones, falling back if they return false. This is already a built-in feature.

**Comment by codeedog** [comment]
@calculuschild however, the extension code handles the regex and scanning and format for tokenization, etc. Specializing just the renderer is a desired behavior. That’s why the regular marked code provides for it. This is a similar situation.

**Comment by codeedog** [comment]
I’ll also add that the override code works great for an extension. It handles multiple layers. It’s just that last call is `false` that’s the problem.

**Comment by calculuschild** [comment]
If you want two different "renderer" behaviors, you can build them as follows:

1) Design the first extension with the default behavior. This can include the Tokenizer as well.
2) Design a second extension but exclude the `tokenizer` field, so it only has Renderer behavior. Have it return false if you want to fall back to the first extension.
3) Load both extensions : `marked.use({ extensions: [myDefaultExtenions, myOverridingRendererExtension] });`

**Comment by calculuschild** [comment]
> It’s just that last call is false that’s the problem.

Yep, since you are trying to override a built-in renderer that doesn't actually exist (this.renderer.media is not built-in to Marked.js), when it finally falls all the way back there is nothing there.

**Comment by codeedog** [comment]
> when it finally falls all the way back there is nothing there.

that’s not quite what the code in marked is doing. The last call returns false and the code looks for “media” on Renderer, which doesn’t exist, but assumes it does. If it just didn’t do that (notice the undefined and not call it), and instead returned false, it’d be fine.

**Comment by calculuschild** [comment]
That may be, but this is still doing something in a way that Marked.js isn't designed to function. It is almost certainly better to use the extension feature built specifically for this purpose. Otherwise there may be risk of your code suddenly breaking when an unrelated update occurs. Ala [XKCD](https://xkcd.com/1172/)

**Comment by codeedog** [comment]
I’m out rn, and will check the extension behavior works as you describe when I’m back at my desk. I agree it’s more important to use published APIs. That said, it’d be great to use the override because it works except for that edge case.

**Comment by UziTech** [comment]
> @UziTech may have some insight if this is something we would actually support in the future though.

My plan is to eventually remove `tokenizer` and `renderer` options and just use `extensions` for everything. I'm not sure when/if that will actually happen but this issue shows the confusion of having multiple ways to do the same thing.

**Comment by codeedog** [comment]
Thank you for your patient explanations. I modified the code above to use the `extensions` property and it worked as expected. I've left a snippet of the modifications below in case anyone comes here to read this.

I also went back and re-read the documentation surrounding extensions and although the docs do suggest this methodology or at least allow for it, it's quite buried considering the alternatives that are documented. Perhaps a few more lines showing the recommended override behavior for extensions would clear up any future misunderstandings. Also, it would seem that extending in this manner means one receives a `token` and not the function signatures typical of the current `renderer` model. I don't think this is a problem, just something that ought to be highlighted in the explanation.

````
// Receive token, grab members
function firstFunc(token) {
  const {alt, src} = token;
  console.log("Func1:", renderInFirstFunc);
  return renderInFirstFunc && `<ff>${src}, ${alt}</ff>`;
}

// Succeeds
marked.use({ extensions: [{ name: 'media', renderer: firstFunc }]});
marked.use({ extensions: [{ name: 'media', renderer: ({alt, src}) => { console.log("Func2"); return false; }}]});
````
